### PR TITLE
Support security update repository to fetch source package from it

### PR DESCRIPTION
**Pull request by masami256**
This PR support fetch updated package from debian security update repository(http://security.debian.org/debian-security/pool).

# Motivation

Currently, when package is updated by security issue, we need to wait a point release. However, it'd be nice to get updated package from security update repository that we can get security update quickly.

## basic design

Basically following lines are required in conf/local.conf.

```
DEBIAN_SOURCE_ENABLED = "1"
DEBIAN_SRC_FORCE_REGEN = "1"

DEBIAN_SECURITY_UPDATE_ENABLED = "1"
DEBIAN_SECURITY_UPDATE_MIRROR = "http://security.debian.org/debian-security/pool"
MIRRORS+="${DEBIAN_MIRROR}    ${DEBIAN_SECURITY_UPDATE_MIRROR}"
```

To enable/disable security update repository, DEBIAN_SECURITY_UPDATE_ENABLED variable is used.

When DEBIAN_SECURITY_UPDATE_ENABLED is set, debian-source.class checks target repositories by following order.
 
1. security update repository
2. debian mirror server

And, 2 rules. 

1. If new version is exists in security update repository, set DEBIAN_SECURITY_UPDATE_MIRROR as mirror server 
2. If same version exists in both repositories,  set DEBIAN_MIRROR as mirror server. 

Why "if same version exists in both repositories,  set DEBIAN_MIRROR as mirror server." is that it supports disable security update repository when user don't need it after a debian point release.

## debian package version check

T0 support security update repository, it needs to compare debian package versions.
I adopt python3-debian package to do it. This package is useful to compare debian package versions.
So I added this package to Dockerfile.

## Update test scripts

tests/build_test and tests/qemu_ptest scripts are able to enable security update repository. If TEST_ENABLE_SECURITY_UPDATE environment variable is set to 1, security update repository will be enabled.

## Test this PR

Debian 10.12 was release recently, there is no new security update packages as of 2022/03/30. Therefore, there is no way to test this PR straightforward.

One of idea is only check debian security update repo like following patch.

```
diff --git a/classes/debian-source.bbclass b/classes/debian-source.bbclass
index 45472840..da2139a9 100644
--- a/classes/debian-source.bbclass
+++ b/classes/debian-source.bbclass
@@ -264,7 +264,7 @@ python debian_source_eventhandler() {
     if debian_source_enabled == '0':
         # Nothing to do
         return
-
+    update_targets.pop(1)
     for update_target in update_targets:
         if not fetch_Sources_xz(update_target['debian_mirror'], update_target['debian_security_update_enabled'], d):
             continue
```

#  example test results

## Update *.inc files

Enter docker container via following command then update .inc files.

```
$ make -C meta-debian/docker

~~~

deby@c78a928c7243:~/build$ echo 'DEBIAN_SOURCE_ENABLED="1"' >> conf/local.conf 
deby@c78a928c7243:~/build$ echo 'DEBIAN_SRC_FORCE_REGEN="1"' >> conf/local.conf 
deby@c78a928c7243:~/build$ echo 'DEBIAN_SECURITY_UPDATE_ENABLED="1"' >> conf/local.conf 
deby@c78a928c7243:~/build$ echo 'DEBIAN_SECURITY_UPDATE_MIRROR="http://security.debian.org/debian-security/pool"' >> conf/local.conf 
deby@c78a928c7243:~/build$ echo 'MIRRORS+="${DEBIAN_MIRROR}    ${DEBIAN_SECURITY_UPDATE_MIRROR}"' >> conf/local.conf 
deby@c78a928c7243:~/build$ bitbake openssl -c fetch
WARNING: debian security update repository is enabled
Checking Debian Release ...
Fetching Debian Sources.xz ...
Parsing Debian Sources.xz ...
WARNING: Source code for package "aide" has been updated from "0.16.1-1" to "0.16.1-1+deb10u1"
WARNING: Source code for package "nss" has been updated from "3.42.1-1+deb10u3" to "3.42.1-1+deb10u5"
WARNING: Source code for package "expat" has been updated from "2.2.6-2+deb10u1" to "2.2.6-2+deb10u4"
WARNING: Source code for package "openssl" has been updated from "1.1.1d-0+deb10u7" to "1.1.1d-0+deb10u8"
Parsing recipes: 100% |################################################################################################################################################################| Time: 0:00:23
Parsing of 1041 .bb files complete (0 cached, 1041 parsed). 1821 targets, 229 skipped, 0 masked, 0 errors.
NOTE: Resolving any missing task queue dependencies

Build Configuration:
BB_VERSION           = "1.42.0"
BUILD_SYS            = "x86_64-linux"
NATIVELSBSTRING      = "universal"
TARGET_SYS           = "aarch64-deby-linux"
MACHINE              = "qemuarm64"
DISTRO               = "deby-tiny"
DISTRO_VERSION       = "10.0"
TUNE_FEATURES        = "aarch64 armv8a crc"
TARGET_FPU           = ""
meta                 
meta-poky            = "warrior:d4b57c68b22027c2bedff335dee06af963e4f8a8"
meta-debian          = "support-secuirty-update-repo:437a550169be1000fe0253278c8db65de4677071"

Initialising tasks: 100% |#############################################################################################################################################################| Time: 0:00:00
NOTE: Executing RunQueue Tasks
NOTE: Tasks Summary: Attempted 1 tasks of which 0 didn't need to be rerun and all succeeded.

Summary: There were 5 WARNING messages shown.
```

## build with build_test script

```
masami@debian10:~/meta-debian-dev/meta-debian$ env | grep TEST_
TEST_MACHINES=qemuarm64
TEST_DISTROS=deby-tiny
TEST_ENABLE_SECURITY_UPDATE=1
TEST_PACKAGES=openssl


masami@debian10:~/meta-debian-dev/meta-debian/tests$ make -C ../docker build_test
make: Entering directory '/home/masami/meta-debian-dev/meta-debian/docker'
docker-compose run --rm build_test
WARNING: The PTEST_RUNNER_TIMEOUT variable is not set. Defaulting to a blank string.
WARNING: The http_proxy variable is not set. Defaulting to a blank string.
WARNING: The https_proxy variable is not set. Defaulting to a blank string.
WARNING: The no_proxy variable is not set. Defaulting to a blank string.
WARNING: The TEST_DISTRO_FEATURES variable is not set. Defaulting to a blank string.
Creating docker_build_test_run ... done
NOTE: Setup build directory.
You had no conf/local.conf file. This configuration file has therefore been
created for you with some default values. You may wish to edit it to, for
example, select a different MACHINE (target hardware). See conf/local.conf
for more information as common configuration options are commented.

You had no conf/bblayers.conf file. This configuration file has therefore been
created for you with some default values. To add additional metadata layers
into your configuration please add entries to conf/bblayers.conf.

The Yocto Project has extensive documentation about OE including a reference
manual which can be found at:
    http://yoctoproject.org/documentation

For more information about OpenEmbedded see their website:
    http://www.openembedded.org/

NOTE: Testing distro deby-tiny ...
NOTE: Testing machine qemuarm64 ...
NOTE: These recipes will be tested: openssl
NOTE: Building openssl ...
NOTE: Build openssl: PASS
make: Leaving directory '/home/masami/meta-debian-dev/meta-debian/docker'
```

## Run qemu test

```
masami@debian10:~/meta-debian-dev/meta-debian/tests$ make -C ../docker qemu_ptest
make: Entering directory '/home/masami/meta-debian-dev/meta-debian/docker'                                                                                                                            
docker-compose run --rm qemu_ptest
WARNING: The http_proxy variable is not set. Defaulting to a blank string.
WARNING: The https_proxy variable is not set. Defaulting to a blank string.
WARNING: The no_proxy variable is not set. Defaulting to a blank string.
WARNING: The TEST_DISTRO_FEATURES variable is not set. Defaulting to a blank string.
Creating docker_qemu_ptest_run ... done
mkstemp: No such file or directory
NOTE: Setup build directory.
You had no conf/local.conf file. This configuration file has therefore been
created for you with some default values. You may wish to edit it to, for
example, select a different MACHINE (target hardware). See conf/local.conf
for more information as common configuration options are commented.
                                                                                                                                                                                                      
You had no conf/bblayers.conf file. This configuration file has therefore been
created for you with some default values. To add additional metadata layers
into your configuration please add entries to conf/bblayers.conf.

The Yocto Project has extensive documentation about OE including a reference
manual which can be found at:
    http://yoctoproject.org/documentation

For more information about OpenEmbedded see their website:
    http://www.openembedded.org/

NOTE: These packages will be tested: openssl
NOTE: Testing distro deby-tiny ...
NOTE: Testing machine qemuarm64 ...
Parsing recipes: 100% |################################################################################################################################################################| Time: 0:00:36
Parsing of 1041 .bb files complete (0 cached, 1041 parsed). 1821 targets, 223 skipped, 0 masked, 0 errors.
NOTE: Resolving any missing task queue dependencies

Build Configuration:
BB_VERSION           = "1.42.0"
BUILD_SYS            = "x86_64-linux"
NATIVELSBSTRING      = "debian-10"
TARGET_SYS           = "aarch64-deby-linux"
MACHINE              = "qemuarm64"
DISTRO               = "deby-tiny"
DISTRO_VERSION       = "10.0"
TUNE_FEATURES        = "aarch64 armv8a crc"
TARGET_FPU           = ""
meta                 
meta-poky            = "warrior:d4b57c68b22027c2bedff335dee06af963e4f8a8"
meta-debian          = "support-secuirty-update-repo:f3cc114948a5e99012fcafe6c6c3c927e8399b6e"
Initialising tasks: 100% |#############################################################################################################################################################| Time: 0:00:00
Sstate summary: Wanted 426 Found 0 Missed 426 Current 0 (0% match, 0% complete)
NOTE: Executing SetScene Tasks
NOTE: Executing RunQueue Tasks
NOTE: Tasks Summary: Attempted 1720 tasks of which 5 didn't need to be rerun and all succeeded.
nohup: appending output to 'nohup.out'
NOTE: Waiting for SSH to be ready... (5s / 60s)
NOTE: Waiting for SSH to be ready... (10s / 60s)
mesg: not a tty
Running ptest for openssl ...
openssl: PASS/SKIP/FAIL = 147/8/0
mesg: not a tty
make: Leaving directory '/home/masami/meta-debian-dev/meta-debian/docker'
```

I needed to set following variables to wait ptest was finished.

```
COMPOSE_HTTP_TIMEOUT=4000
PTEST_RUNNER_TIMEOUT=4000
```

### CODE REVIEW

File: scripts/install-deps.sh [hunk 1]
```diff
@@ -14,7 +14,7 @@ PKG_ESSENTIALS_DEBIAN=" \
 	gawk wget git-core diffstat unzip texinfo gcc-multilib \
 	build-essential chrpath socat cpio python python3 python3-pip \
 	python3-pexpect xz-utils debianutils iputils-ping python3-git \
-	python3-jinja2 libegl1-mesa libsdl1.2-dev xterm \
+	python3-jinja2 libegl1-mesa libsdl1.2-dev xterm pytnon3-debian \
```
[the remark this answers is not in the document]
Comment (masami256) [reply scripts/install-deps.sh]: @aliceinwire thanks! fixed.

**Comment by masami256** [comment]
@zuka0828 It's sufficient for me. I can also keep the partial updates only in my own branch.
By the way, I think there will not be point releases after the Debian LTS period begins, but do you have any plans to pull security updates to the meta-debian at that time?

**Comment by zuka0828** [comment]
> but do you have any plans to pull security updates to the meta-debian at that time?

Yes, and I think this activity requires the feature added by this PR.
According to LTS document[0], both the main and security repositories are required in `sources.list` to use LTS packages, but actually it seems that all LTS packages are distributed through the security repository[1] and the main repository[2] is not updated by LTS team[3].

The timing when meta-debian recipes should be updated has not been decided.
May need to provide simple periodic updates (e.g. per a month) because no point release in LTS as you pointed out.

* [0] https://wiki.debian.org/LTS/Using
* [1] https://security.debian.org/debian-security/dists/stretch/updates/main/
* [2] http://ftp.debian.org/debian/dists/stretch/main/
* [3] https://www.linux-magazine.com/Issues/2017/194/Debian-LTS

File: doc/6.security-update.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+#### Enable security update repository
+
+Please set following lines in your conf/local.conf
+
+```
+DEBIAN_SOURCE_ENABLED = "1"
+DEBIAN_SRC_FORCE_REGEN = "1"
+
+DEBIAN_SECURITY_UPDATE_ENABLED = "1"
+DEBIAN_SECURITY_UPDATE_MIRROR = "http://security.debian.org/debian-security/pool"
```
[the remark this answers is not in the document]
Comment (masami256) [reply doc/6.security-update.md]: @zuka0828 Thank you for the review!  I rebased to latest warrior branch and  changed to include /updates patch in DEBIAN_SECURITY_UPDATE_MIRROR.

It's my test result.

Before run update, there is zlib_1.2.11.dfsg-1.dsc file in download directory.

```
masami@debian10:~/metadebian/build$ ls -la downloads/zlib*
-rw-r--r-- 1 masami masami 18956 Apr 28  2018 downloads/zlib_1.2.11.dfsg-1.debian.tar.xz
-rw-r--r-- 1 masami masami   141 Apr  8 13:25 downloads/zlib_1.2.11.dfsg-1.debian.tar.xz.done
-rw-r--r-- 1 masami masami  2266 Apr 28  2018 downloads/zlib_1.2.11.dfsg-1.dsc
-rw-r--r-- 1 masami masami   141 Apr  8 13:25 downloads/zlib_1.2.11.dfsg-1.dsc.done
```

Run fetch task that causes zlib package update.

```
masami@debian10:~/metadebian/build$ bitbake -f zlib -c fetch
Loading cache: 100% |##########################################################################################################################################################################################################| Time: 0:00:00
Loaded 1818 entries from dependency cache.
WARNING: debian security update repository is enabled
Checking Debian Release ...
Fetching Debian Sources.xz ...
Parsing Debian Sources.xz ...
WARNING: Source code for package "zlib" has been updated from "1.2.11.dfsg-1" to "1.2.11.dfsg-1+deb10u1"
Checking Debian Release ...
Fetching Debian Sources.xz ...
Parsing Debian Sources.xz ...
Parsing recipes: 100% |########################################################################################################################################################################################################| Time: 0:00:07
Parsing of 1041 .bb files complete (782 cached, 259 parsed). 1821 targets, 223 skipped, 0 masked, 0 errors.
NOTE: Resolving any missing task queue dependencies

Build Configuration:
BB_VERSION           = "1.42.0"
BUILD_SYS            = "x86_64-linux"
NATIVELSBSTRING      = "universal"
TARGET_SYS           = "aarch64-deby-linux"
MACHINE              = "qemuarm64"
DISTRO               = "deby-tiny"
DISTRO_VERSION       = "10.0"
TUNE_FEATURES        = "aarch64 armv8a crc"
TARGET_FPU           = ""
meta                 
meta-poky            = "warrior:d4b57c68b22027c2bedff335dee06af963e4f8a8"
meta-debian          = "support-secuirty-update-repo:4ba23d99f691c71770ab42d8724a0f62b678c133"

NOTE: Tainting hash to force rebuild of task /home/masami/metadebian/poky/meta-debian/recipes-debian/zlib/zlib_debian.bb, do_fetch                                                                                             | ETA:  0:00:00
WARNING: /home/masami/metadebian/poky/meta-debian/recipes-debian/zlib/zlib_debian.bb.do_fetch is tainted from a forced run                                                                                                     | ETA:  0:00:00
Initialising tasks: 100% |#####################################################################################################################################################################################################| Time: 0:00:00
NOTE: Executing RunQueue Tasks
NOTE: Tasks Summary: Attempted 1 tasks of which 0 didn't need to be rerun and all succeeded.

Summary: There were 3 WARNING messages shown.
```
Then, I got new version of zlib package. 
```
masami@debian10:~/metadebian/build$ ls -la downloads/zlib*
-rw-r--r-- 1 masami masami  23092 Mar 29 00:12 downloads/zlib_1.2.11.dfsg-1+deb10u1.debian.tar.xz
-rw-r--r-- 1 masami masami    141 Apr 12 09:59 downloads/zlib_1.2.11.dfsg-1+deb10u1.debian.tar.xz.done
-rw-r--r-- 1 masami masami   2774 Mar 29 00:12 downloads/zlib_1.2.11.dfsg-1+deb10u1.dsc
-rw-r--r-- 1 masami masami    141 Apr 12 09:59 downloads/zlib_1.2.11.dfsg-1+deb10u1.dsc.done
-rw-r--r-- 1 masami masami  18956 Apr 28  2018 downloads/zlib_1.2.11.dfsg-1.debian.tar.xz
-rw-r--r-- 1 masami masami    141 Apr  8 13:25 downloads/zlib_1.2.11.dfsg-1.debian.tar.xz.done
-rw-r--r-- 1 masami masami   2266 Apr 28  2018 downloads/zlib_1.2.11.dfsg-1.dsc
-rw-r--r-- 1 masami masami    141 Apr  8 13:25 downloads/zlib_1.2.11.dfsg-1.dsc.done
-rw-r--r-- 1 masami masami 370248 Sep  2  2019 downloads/zlib_1.2.11.dfsg.orig.tar.gz
-rw-r--r-- 1 masami masami    141 Apr 12 09:59 downloads/zlib_1.2.11.dfsg.orig.tar.gz.done
```

zlib.inc file structure is following.

```
masami@debian10:~/metadebian/poky/meta-debian/recipes-debian/sources$ cat zlib.inc 
# This is generated by debian-source.bbclass
DPV = "1.2.11.dfsg-1+deb10u1"
DPV_EPOCH = "1"
REPACK_PV = "1.2.11.dfsg"
PV = "1.2.11"

DEBIAN_SRC_URI = " \
    ${DEBIAN_SECURITY_UPDATE_MIRROR}/main/z/zlib/zlib_1.2.11.dfsg-1+deb10u1.dsc;name=zlib_1.2.11.dfsg-1+deb10u1.dsc \
    ${DEBIAN_SECURITY_UPDATE_MIRROR}/main/z/zlib/zlib_1.2.11.dfsg.orig.tar.gz;name=zlib_1.2.11.dfsg.orig.tar.gz \
    ${DEBIAN_SECURITY_UPDATE_MIRROR}/main/z/zlib/zlib_1.2.11.dfsg-1+deb10u1.debian.tar.xz;name=zlib_1.2.11.dfsg-1+deb10u1.debian.tar.xz \
"

SRC_URI[zlib_1.2.11.dfsg-1+deb10u1.dsc.md5sum] = "db557e73a35459dcf470e18d7fc63acc"
SRC_URI[zlib_1.2.11.dfsg.orig.tar.gz.md5sum] = "2950b229ed4a5e556ad6581580e4ab2c"
SRC_URI[zlib_1.2.11.dfsg-1+deb10u1.debian.tar.xz.md5sum] = "82dac8e2b7814db97e11f5675a6bcdc1"

SRC_URI[zlib_1.2.11.dfsg-1+deb10u1.dsc.sha256sum] = "3eddd036f314f8a7717c41fe7d3c935653afeb6b7f7bf6d9ad80c10ef1c4bdc3"
SRC_URI[zlib_1.2.11.dfsg.orig.tar.gz.sha256sum] = "80c481411a4fe8463aeb8270149a0e80bb9eaf7da44132b6e16f2b5af01bc899"
SRC_URI[zlib_1.2.11.dfsg-1+deb10u1.debian.tar.xz.sha256sum] = "eb26660e5b8a39f945a4fe1284e29b0279ded3513327e3cbd51c51921758f13f"
```

**Comment by masami256** [comment]
@zuka0828 Thank you for the description. 

> Yes, and I think this activity requires the feature added by this PR.

I hope this PR' feature helps in the future.
 
>  May need to provide simple periodic updates (e.g. per a month) because no point release in LTS as you pointed out.

It sounds good to me.

File: doc/6.security-update.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (zuka0828) [reply doc/6.security-update.md]: @masami256 Thank you very much for the quick update

**Comment by zuka0828** [comment]
@masami256 Confirmed that the change does not affect the normal build without `DEBIAN_SECURITY_UPDATE_*`, and the tests for `DEBIAN_SECURITY_UPDATE_*` you explained passed. Thanks a lot.

**Comment by masami256** [comment]
@zuka0828 Thank you very much!

**Pull request merged**